### ISSUE TITLE
Incorrect result caused by ORDER BY clause

### ISSUE THREAD

**Issue by Aurora-yzj** [FalkorDB/FalkorDB#776]
Hi! I discovered a bug in FalkorDB that `order by` clause at some point could cause the database to wrongly produce empty results. Steps to reproduce are as follows. 
1. Set the initial data in the DB.
```cypher
CREATE (n0 :L2{k11 : "Gat10", id : 1, k13 : false, k15 : 933871164, k14 : "cWtEEOW5X"})
CREATE (n0 :L3:L4{k17 : false, k16 : true, k18 : true, k20 : "EyDIvgH", k22 : "yweed", k21 : 2048372834, id : 4, k24 : true, k26 : 972481772, k25 : "mjEb20"})
CREATE (n0 :L4{k22 : "MlVlwZl2", id : 6, k24 : false, k23 : 1920403648, k26 : -1490105746, k25 : "YveOk3"})
CREATE (n0 :L4{k22 : "nZLrxiNW", id : 9, k24 : true, k23 : -14262776, k26 : -2021157956, k25 : "eY2SNZM0O"})
MATCH (n0 {id : 1}), (n1 {id : 6}) MERGE(n0)-[r :T7{k80 : false, k75 : "MhUsMP7b", k77 : -1027999688, k76 : "DvXbdJ", id : 11, k79 : true, k78 : true}]->(n1)
MATCH (n0 {id : 4}), (n1 {id : 9}) MERGE(n0)-[r :T9{k91 : false, k90 : true, k92 : false, k88 : 1567379845, k87 : true, id : 17, k89 : -867329166}]->(n1)
```
2. Consider the following query:
```cypher
MATCH (n0)<-[r0]-(n1{id:4}), (n2)-[r1]->(n0), (n3)<-[r2]-(n2) WHERE true WITH DISTINCT n1, (n1.k22) AS a7, n0 ORDER BY (n1.k16), (n1.k18), (n1.k20) DESC, (n1.k21), (n1.id) DESC, (n1.k22), (n1.k24), (n1.k25), (n1.k26), (n1.id), a7 DESC SKIP 0 WHERE ((n0.k24) = true) MATCH (n4)<-[r3]-(n2), (n1)-[r4]->(n0) WHERE ((((n4.k2)+(n1.k22+((r4.k111)+(n2.k22)))) CONTAINS ('y'+'weedVAbRUfFA')) OR (((n2.k24) = true) OR (((n1.k17) = true) OR (-116 <= (r4.k115))))) RETURN ((n0.k24) = true)
```
It gives an empty result. However, if we remove the order by clause in WITH clause:
```cypher
MATCH (n0)<-[r0]-(n1{id:4}), (n2)-[r1]->(n0), (n3)<-[r2]-(n2) WHERE true WITH DISTINCT n1, (n1.k22) AS a7, n0  WHERE ((n0.k24) = true) MATCH (n4)<-[r3]-(n2), (n1)-[r4]->(n0) WHERE ((((n4.k2)+(n1.k22+((r4.k111)+(n2.k22)))) CONTAINS ('y'+'weedVAbRUfFA')) OR (((n2.k24) = true) OR (((n1.k17) = true) OR (-116 <= (r4.k115))))) RETURN ((n0.k24) = true)
```
It normally gives the result.

Thank you!

**Comment by swilly22** [comment]
Hi @Aurora-yzj, thank you for all of your recent filings!
I'll assume you're catching all of these bugs using some sort of a fuzzer ?
Would you mind sharing some information on the method you're using ?

**Comment by Aurora-yzj** [comment]
> Hi @Aurora-yzj, thank you for all of your recent filings! I'll assume you're catching all of these bugs using some sort of a fuzzer ? Would you mind sharing some information on the method you're using ?

Yes! I am trying out a different tactic for finding logic bugs in database systems other than metamorphic-based oracles. My developed framework is currently under experiment, and indeed found some issues (which were actually introduced a long time ago from RedisGraph). I will try my best to minimize the test case and try to locate the root cause, and hope these findings can help you!

**Comment by DvirDukhan** [comment]
Verified against both engines and closing as fixed by the Rust migration.

**Reproduction**

```cypher
CREATE (n0 :L2{k11 : "Gat10", id : 1, k13 : false, k15 : 933871164, k14 : "cWtEEOW5X"})

CREATE (n0 :L3:L4{k17 : false, k16 : true, k18 : true, k20 : "EyDIvgH", k22 : "yweed", k21 : 2048372834, id : 4, k24 : true, k26 : 972481772, k25 : "mjEb20"})

CREATE (n0 :L4{k22 : "MlVlwZl2", id : 6, k24 : false, k23 : 1920403648, k26 : -1490105746, k25 : "YveOk3"})

CREATE (n0 :L4{k22 : "nZLrxiNW", id : 9, k24 : true, k23 : -14262776, k26 : -2021157956, k25 : "eY2SNZM0O"})

MATCH (n0 {id : 1}), (n1 {id : 6}) MERGE(n0)-[r :T7{k80 : false, k75 : "MhUsMP7b", k77 : -1027999688, k76 : "DvXbdJ", id : 11, k79 : true, k78 : true}]->(n1)

MATCH (n0 {id : 4}), (n1 {id : 9}) MERGE(n0)-[r :T9{k91 : false, k90 : true, k92 : false, k88 : 1567379845, k87 : true, id : 17, k89 : -867329166}]->(n1)

MATCH (n0)<-[r0]-(n1{id:4}), (n2)-[r1]->(n0), (n3)<-[r2]-(n2) WHERE true WITH DISTINCT n1, (n1.k22) AS a7, n0 ORDER BY (n1.k16), (n1.k18), (n1.k20) DESC, (n1.k21), (n1.id) DESC, (n1.k22), (n1.k24), (n1.k25), (n1.k26), (n1.id), a7 DESC SKIP 0 WHERE ((n0.k24) = true) MATCH (n4)<-[r3]-(n2), (n1)-[r4]->(n0) WHERE ((((n4.k2)+(n1.k22+((r4.k111)+(n2.k22)))) CONTAINS ('y'+'weedVAbRUfFA')) OR (((n2.k24) = true) OR (((n1.k17) = true) OR (-116 <= (r4.k115))))) RETURN ((n0.k24) = true)

MATCH (n0)<-[r0]-(n1{id:4}), (n2)-[r1]->(n0), (n3)<-[r2]-(n2) WHERE true WITH DISTINCT n1, (n1.k22) AS a7, n0  WHERE ((n0.k24) = true) MATCH (n4)<-[r3]-(n2), (n1)-[r4]->(n0) WHERE ((((n4.k2)+(n1.k22+((r4.k111)+(n2.k22)))) CONTAINS ('y'+'weedVAbRUfFA')) OR (((n2.k24) = true) OR (((n1.k17) = true) OR (-116 <= (r4.k115))))) RETURN ((n0.k24) = true)

MATCH (n0)<-[r0]-(n1 {id:4}), (n2)-[r1]->(n0) RETURN n0.id, n2.id, r0.id, r1.id
```

| Engine | Result |
| --- | --- |
| `falkordb/falkordb:latest` (v4.20.1, C engine) | Both the ORDER-BY and no-ORDER-BY variants from the issue now return the SAME (non-empty) result on v4.20.1 -- `true` (1 row) -- so the literal 'ORDER BY flips empty/non-empty' symptom is gone. But that shared answer is itself wrong: minimal isolation (`MATCH (n0)<-[r0]-(n1{id:4}), (n2)-[r1]->(n0) RETURN n0.id,n2.id,r0.id,r1.id` on a 2-node/1-edge graph) shows latest binds the SAME physical relationship to both r0 and r1 (r0.id=17 and r1.id=17), violating Cypher's relationship-uniqueness rule across comma-separated patterns within one MATCH -- with only one edge (4->9) in the graph, no valid binding exists where r0 != r1, so the correct result is 0 rows, not 1. |
| `falkordb/falkordb:edge-rs` (Rust engine) | Both variants return an empty result (0 rows) on edge-rs -- correct, matching ground truth. The minimal isolated repro also correctly returns 0 rows: edge-rs refuses to bind r0 and r1 to the same relationship. |

The issue's own framing (ORDER BY causes an incorrect EMPTY result; removing it gives the 'normal' non-empty result) turned out to be backwards. I derived ground truth by hand: the first MATCH clause requires r0 (an edge FROM n1={id:4} TO n0) and r1 (an edge FROM some n2 TO that same n0) to be DIFFERENT relationships (openCypher relationship-uniqueness applies across all relationship variables declared in one MATCH clause, even across comma-separated sub-patterns). In this dataset the only edge out of id=4 is 4->9, and the only edge into id=9 is that same edge -- so r0 and r1 can only be satisfied by the identical relationship, which is disallowed, and the query must return 0 rows. Today on latest, BOTH variants (with/without ORDER BY) incorrectly return 1 row by letting r0 and r1 alias the same edge -- a relationship-isomorphism bug independent of ORDER BY (confirmed with a standalone 2-node minimal repro). On edge-rs, both variants correctly return 0 rows, and the minimal isolated repro confirms edge-rs properly enforces relationship uniqueness here. So while the *specific* ORDER-BY-triggered inconsistency described in the issue no longer manifests as such on either engine, the actual underlying defect (wrong, non-unique relationship binding producing a phantom row) still reproduces on latest and is genuinely fixed on edge-rs -- verified with named (non-anonymous) relationship variables throughout, so this is not an instance of the known anonymous-relationship-variable isomorphism gap.

The Rust engine became the default development engine in #2279 and ships as the
rolling `edge-rs` / `edge` Docker tags, so this issue is resolved there. Reopen if
you still hit it on `edge-rs`.
